### ISSUE TITLE
Better API or docs

### ISSUE THREAD

**Issue by ai** [postcss/postcss-cli#1]
Let’s think how we make CLI better.

**Comment by corysimmons** [comment]
@hawkrives seemed to have the best idea. @hawkrives care to put your proposed syntax in here?

**Comment by corysimmons** [comment]
@ai If we rewrite this from scratch does this need to be a fork? Looks funny.

**Comment by hawkrives** [comment]
Oh, uh, I'll go look for my proposal. Was it in the old repo somewhere?

**Comment by hawkrives** [comment]
Found it.

TL;DR: `postcss -p autoprefixer input.css > output.css` and `postcss -p [ autoprefixer -b "browsers" ] input.css > output.css`, using [subarg].


For other references, check out Browserify: `browserify -t [ babelify --experimental ] -e input.js > output.js`

Browserify gets all the arguments outside of the square brackets, and passes the ones inside to the invoked command. 

The arguments are parsed into something like:

- Browserify
   - `-e` input.js
   - `-t` 
      - babelify
         - `--experimental`

Substack pulled the subarg parsing into its own package, at [subarg].

An example for PostCSS would probably look something like `postcss -p [ autoprefixer -b "browsers" ] input.css > output.css`

Note that if you don't want to pass any args to the plugin, you dont need the brackets: `postcss -p autoprefixer input.css > output.css`

[subarg]: https://github.com/substack/subarg

**Comment by hawkrives** [comment]
and/or other proposal:

`postcss [-p plugin] [some input files] [-o output-file] [--out-dir dir]`

- Any number of plugins, each with its own `-p` .
- Any number of input files
	- If you give one input file, and no output, it'll write to stdout
	- If you give `-o` with one input file, it writes to that file
	- If you give multiple inputs, and no `--out-dir`, it'll overwrite the input files
	- If you give `--out-dir` with multiple input files, it'll write the corresponding files to the specified directory

I took `--out-dir` from [Babel's cli](https://babeljs.io).

**Comment by corysimmons** [comment]
@hawkrives Thanks. How would multiple plugin options work?

For instance, what if I wanted to pass `browsers: last 2 versions` and `remove: false` with Autoprefixer?

**Comment by hawkrives** [comment]
Probably something like `postcss -p [autoprefixer --browsers="last 2 versions" --no-remove] input.css`

which would result in the autoprefixer plugin getting `{browsers: 'last 2 versions', remove: false}` as options.

(subarg and therefore minimist treat `--no-arg` and `--arg=false` the same, returning `{arg: false}` either way.)

**Comment by corysimmons** [comment]
Gotcha. I think I like the plugin name on the outside of the brackets a bit better.

```
postcss -p [autoprefixer --browsers="last 2 versions" --no-remove] -p [cssnano --safe --sourcemap] -p postcss-cssnext input.css
postcss -p autoprefixer [--browsers="last 2 versions" --no-remove] -p cssnano [--safe --sourcemap] -p postcss-cssnext input.css
```

I'm also not sure I like the arrow.

```
postcss input.css > output.css
postcss input.css -o output.css
```

Thoughts?

**Comment by corysimmons** [comment]
Anyone have any special feelings about https://github.com/sindresorhus/meow ? Looks pretty kewl to me. :100:

**Comment by TrySound** [comment]
@corysimmons meow does not allow to use file config, only field from package.json. We have https://github.com/davidtheclark/cosmiconfig for such cases.

**Comment by corysimmons** [comment]
We can use the two together right?

**Comment by TrySound** [comment]
Hm.. yep, right.

**Comment by corysimmons** [comment]
:100:

**Comment by vincentorback** [comment]
Adding value of `false` to the output option should not write to any file: `-o false`.
This would be great if you just want to parse CSS for eg. linting.

**Comment by vitorgalvao** [comment]
> I'm also not sure I like the arrow.

That’s standard unix redirection, so it makes sense for it to be there, you can’t really take it out.

> Adding value of `false` to the output option should not write to any file: `-o false`.

Why complicate? Just behave like any other *nix tool.

I’m pretty sure what @hawkrives meant was for it to behave akin to `curl`. If you don’t specify an output, it’ll just output to STDOUT (i.e. show on the terminal). If you give it `> file_name`, it’ll write to `file_name` instead of being shown in the terminal. `-o file_name` would behave exactly the same as `> file_name`, but `-o` needs to be built into the tool, while `>` will just work by default, since it’s your shell doing the work, there.

**Comment by davidtheclark** [comment]
Yeah, we need to follow these CLI conventions that allow for piping, redirection, etc.

**Comment by hawkrives** [comment]
@corysimmons: @vitorgalvao has it right. I've been thinking that if you give it one input file, and no `-o` flag, it'll just print the processed file to `stdout`, where you can redirect it with your shell's redirection (usually `>`). This allows the famed unix composability, like `postcss -p autoprefixer input.css | wc -c` (counts the number of characters [or bytes] in the processed file.)

@corysimmons: About the braces: Babel (and babelify) has developed a lot of configurability with its latest release. And I was thinking that we could just use the [subarg](https://github.com/substack/subarg) library to split up the options, which uses the first word inside the square brackets to denote the name of the … argument group, I guess we'd call it.

**Comment by corysimmons** [comment]
Arrow is cool then. :+1: 

https://github.com/substack/subarg/blob/master/index.js seems pretty straightforward and hasn't been updated in years. We could make something similar and change the syntax from `-p [foo options]` to `-p foo [options]`, then fork Meow and use it instead of just standard Minimist. Just a matter of what syntax everyone prefers.

Even if we use the `-p [foo options]` syntax and untouched subarg, we could fork Meow and add subarg instead of just minimist? It provides lots of sugar.

**Comment by corysimmons** [comment]
@hawkrives Whatcha think?

**Comment by ai** [comment]
BTW, we can move current `API.md` to some kind of JSDoc into JS sources.

**Comment by hawkrives** [comment]
wait, did I not reply last night? whoops.

@corysimmons: 

- I'm pretty sure it'll be simpler to parse the options with the plugin name inside of the square brackets.
- I'm not married to using subarg, it's just that it exists already, if that makes sense?

**Comment by hawkrives** [comment]
and minimist is actually pretty powerful by itself: you can do automatic booleans and pass nested objects as args.

(live example: https://tonicdev.com/hawkrives/56b0cc8b65d8fc0c00633ef6)

`-p 1 --no-power --nested.object.key --no-nested.object.value`

```json
{
  "_": [],
  "p": 1,
  "power": false,
  "nested": {
    "object": {
      "key": true,
      "value": false
    }
  }
}
```

**Comment by hawkrives** [comment]
But maybe we should figure out how we want it to look before we go bikeshedding on argument parsers :stuck_out_tongue_winking_eye:

**Comment by corysimmons** [comment]
> But maybe we should figure out how we want it to look before we go bikeshedding on argument parsers

:100: 

I still like `cli -p plugin [--option1=foo --option2=bar] -w in.css > out.css`. Just waiting on a couple other people to chime in.

**Comment by ben-eb** [comment]
I don't like the above, it's ambiguous. command line arguments should be in the form of a key value pair, or a boolean flag. A command line tool should not have to rely on the arguments order for option parsing as that syntax does.

Personally I use minimist which handles nested options well, as @hawkrives wrote. I don't see the point of reinventing the wheel for arguments parsing here, as a good range of solutions for this exist already.

**Comment by corysimmons** [comment]
I'm all about minimist as it makes our job (and my desire to use Meow) that much easier. I'm just personally not a fan of that syntax. I dunno, can someone show me a minimist cmd that might include 2 or 3 postcss plugins with a few options assigned to each one? Like a real world command?

**Comment by hawkrives** [comment]
I'm coming up with some examples; should have them posted within the hour

**Comment by corysimmons** [comment]
haha, no rush :0

**Comment by hawkrives** [comment]
This assumes that we implement something that automatically searches for plugins under the `postcss-` namespace, so that a reference to `color-gray` returns the `postcss-color-gray` plugin.

```bash
postcss \
	--parser postcss-scss \
	-p precss \
	-p color-gray \
	-p rebeccapurple \
	-p [selector-matches --no-line-break] \
	-p [autoprefixer --browsers '> 1%'] \
	-p [short-font-size --prefix 'x'] \
	-p short-position \
	-p [discard-comments --remove-all] \
	-p [short --weight.disable] \
	-p [partial-import --extension=css] \
	input.css -o output.css
```

or, on a single line:

```
postcss --parser postcss-scss -p precss -p color-gray -p rebeccapurple -p [selector-matches --no-line-break] -p [autoprefixer --browsers '> 1%'] -p [short-font-size --prefix 'x'] -p short-position -p [discard-comments --remove-all] input.css -o output.css
```

I should say: I view this as a partial alternative to the JS api, so I don't necessarily intend for people to write these long commands by hand – but this will let them write it once and, say, use it from Make, or a shell script, or anywhere that can run a command on the command line, without needing to whip up a little script.

Anyway, that should exercise most of the api surface, I think.

I'm also figuring that we'll essentially pass the options for each plugin to their exported functions, since most modules accept it there.

> [Load Short as a PostCSS plugin:](https://github.com/jonathantneal/postcss-short#postcss)
> ```js
postcss([
    require('postcss-short')({ /* options */ })
]);
```

---

I'm not entirely sure what I said here, but I don't have time to make any revisions. Hopefully someone can get some meaning out of these words :smile:.

I'm getting my wisdom teeth removed this afternoon, so I might be a bit… absent for some unknown period of time.

Oh, and the `\` in the first example are just there to tell the shell, "ignore this line-break."

**Comment by corysimmons** [comment]
@hawkrives This is for minimist or subarg?

**Comment by hawkrives** [comment]
Subarg.

**Comment by hawkrives** [comment]
And therefore mini mist, i think.

I'm building up another example thing,

**Comment by hawkrives** [comment]
See https://tonicdev.com/hawkrives/56b13b16eefec70d00757ae2

**Comment by corysimmons** [comment]
Okay so @ben-eb prefers minimist. @hawkrives prefers subarg. I prefer a fork of subarg.

@ai Please decide on one of these and we can move forward.

**Comment by corysimmons** [comment]
Actually minimist doesn't look too bad, especially if they're using newlines. I don't really like the idea of trimming `postcss-` off of lib names. Seems to add a bit of confusion.

Minimist on newlines with full lib names:
```
postcss \
    --parser postcss-scss \
    -p postcss-precss \
    -p postcss-selector-matches --no-line-break \
    -p autoprefixer --browsers '> 1%' \
    -p postcss-short-font-size --prefix 'x' \
    -p postcss-short-position \
    -p postcss-partial-import --extension=css \
    input.css -o output.css
```

I also like the `-o` flag rather than `>`. If no output is specified it goes to stdout.

Thoughts on this?

**Comment by ben-eb** [comment]
Ideally we skip this idea about custom nested options parsing and use a file/`package.json` config etc, like babel, eslint, ava, and many others. Packages like subarg and minimist provide what we want *if* people are comfortable configuring their whole PostCSS setup using a single command, and seeing as your idea was about making the CLI more user friendly, I don't think any of these syntaxes is friendlier than a JSON/YAML config. So why implement some confusing syntax that is not standard at all over a config file?

Now, minimist provides us with nested options parsing for free, so if we use minimist then you can have all the config in one command if you so wish. We can document this functionality but we should promote the idea of using `package.json` or a config file instead.

Even with one plugin and two options it is confusing with minimist:

```sh
postcss --autoprefixer.browsers 'last 2 versions' --no-autoprefixer.remove in.css > out.css
```

Why not extract that into a config file? It's much more sane and user friendly than some huge shell command.

```json
{
  "plugins": {
    "autoprefixer": {
      "browsers": "last 2 versions",
      "remove": false
    }
  }
}
```

Then run with:

```sh
postcss in.css > out.css
```

Isn't that much simpler?

**Comment by corysimmons** [comment]
> Ideally we skip this idea about custom nested options parsing and use a file/package.json config etc, like babel, eslint, ava, and many others. Packages like subarg and minimist provide what we want if people are comfortable configuring their whole PostCSS setup using a single command, and seeing as your idea was about making the CLI more user friendly, I don't think any of these syntaxes is friendlier than a JSON/YAML config.

I see what you're saying, and I think that will/should be the typical use of it, but you're assuming everyone using this will have a `package.json` file. What about people who want to use the tool globally? I know this is typically a bad practice, but I find myself frequently using Stylus globally to quickly test out something (rather than having to setup a package.json). Those tools you mentioned all provide a global way to use them and specify options. Stylus lets you use plugins globally. I use `stylus -u autoprefixer-stylus -w style.styl` almost daily for testing out stuff.

> So why implement some confusing syntax that is not standard at all over a config file?

What about our CLI syntax wouldn't be standard? Isn't that the point of this issue/discussion?

> Now, minimist provides us with nested options parsing for free, so if we use minimist then you can have all the config in one command if you so wish. We can document this functionality but we should promote the idea of using package.json or a config file instead.

I'm totally fine with promoting the merits of using `package.json` configuration over a huge single line command.

> Even with one plugin and two options it is confusing with minimist...

Is that the correct syntax? Wouldn't it be?
`postcss -p autoprefixer --browsers 'last 2 versions' --no-remove in.css > out.css`

Anyway, I agree with your desire to get users to use config scripts, but I don't want to write a config script every time I want to play around with PostCSS.

**Comment by ben-eb** [comment]
Hmm, ok. Perhaps for global use cases like you mentioned it's fine to have parameters like `--plugin` in the command.

I mean, I'd rather see an implementation that uses either minimist's or subarg's nested argument parsing rather than a custom one. No need to reinvent the wheel, & I personally feel that CLI options should be either key value pairs or booleans. :smile:

---

This example;

```sh
postcss -p autoprefixer --browsers 'last 2 versions' --no-remove in.css > out.css
```

looks like it should expand to this:

```json
{
  "plugins": {
    "autoprefixer": {
      "browsers": "last 2 versions"
    }
  },
  "remove": false
}
```

and;

```sh
postcss -p autoprefixer --no-remove --browsers 'last 2 versions' in.css > out.css
```

looks like it should expand to this:

```json
{
  "plugins": {
    "autoprefixer": {
      "remove": false
    }
  },
  "browsers": "last 2 versions"
}
```

**Comment by corysimmons** [comment]
Alright! That looks good to me @ben-eb 

Let's wait til @hawkrives gets back, since he's put so much thought into this, before moving forward.

**Comment by hawkrives** [comment]
I should be able to post a refined set of ideas tomorrow morning. I'll include:

- example pure CLI commands
- config file ideas (and potential modules)
- … I had a third thing.

Anyway. It's about bedtime for me, so I'll

**Comment by hawkrives** [comment]
gah

**Comment by corysimmons** [comment]
@hawkrives Sounds good. No rush

**Comment by hawkrives** [comment]
Oh, here's a thing: what are some plugins that people actually use? I haven't really ventured beyond autoprefixer in my daily use.

**Comment by corysimmons** [comment]
^ great question. I'm so stuck in my own little world that I just use Lost and Autoprefixer 99% of the time.

- https://github.com/peterramsing/lost#global-grid-settings
- https://github.com/postcss/autoprefixer#browsers
- https://github.com/postcss/postcss-calc#options
- http://cssnext.io/usage
- http://cssnano.co/options/
- https://github.com/postcss/postcss-import#options

Should get you started. :)

Do all PostCSS plugins expose options the same way?

**Comment by ben-eb** [comment]
As far as I am aware, all plugins are configurable by options objects.

**Comment by davidtheclark** [comment]
My two cents: I worry that trying *not* relying on a configuration file, and instead trying to allow plugin configuration via CLI arguments, is a recipe for bugs and headaches. I worry that the `package.json` property idea is the same, too. 

The main problem is that both CLI arguments and `package.json` (or YAML or JSON) configs can only use strings as arguments. But we can't control the APIs of plugins, and some plugins might want functions or regular expressions, or call for multiple arguments instead of a single configuration object, etc. The only setup that allow for the flexibility we need would be a CommonJS module, right? —— like `webpack.config.js`?

If we don't restrict configuration to a CommonJS module, I worry that the CLI project will be endlessly plagued with issues about how "I can't use option x with plugin y in the CLI!" —— And there's nothing we could *do* about those: we couldn't fix them, because we don't have control over the plugins (and the plugins themselves aren't doing anything wrong); we couldn't make a definitive list of options of plugins to avoid, because that would be ridiculous. Instead, we'd have to say "Yeah, I guess the CLI only *mostly* works.* — which is a crappy answer.

So my suggestion would be to follow the pattern of webpack:

- Use a CommonJS module for configuration.
- Including plugins as an array (just like when you currently use postcss in any of its runners).
- Save CLI arguments for PostCSS options and CLI concerns, like `watch`.

Or does somebody have another way to solve the plugin-options-that-aren't-strings problem?

**Comment by ben-eb** [comment]
^ Worth mentioning that JSON supports arrays, booleans and integer options as well as strings. I think for *most* (~80%?) of plugins, these types are sufficient. Personally, have only used a function option in 2 plugins.

https://github.com/ben-eb/postcss-discard-font-face#filter
https://github.com/ben-eb/postcss-discard-comments#removefunction

**Comment by davidtheclark** [comment]
@ben-eb : Right, changed "string" above to be a little more accurate. I agree that for most plugins they're probably sufficient. But what if they're not for 10% of plugins? That would make for a bad CLI UX and a bad time with maintenance, right?

**Comment by ben-eb** [comment]
Having the option to use CommonJS config *as well as* reading config from `package.json` or a file is consistent with other tools (e.g. eslint), so I don't see a problem with that. We need clear documentation so that people are aware that some plugins that use functions as arguments will only work with CJS.

**Comment by davidtheclark** [comment]
Yes, could be done (and cosmiconfig already handles that). I just think it's significantly more awkward for user experience, developer experience, and maintenance to have that kind of *partial* support for *most* plugins if you configure in *some* ways. And the alternative is to provide a single means of configuration that will *always* work for *all* plugins. 

I don't think tools like ESLint are equivalent, because they don't have the requirements that are causing this problem in the first place, right? I think webpack is, and webpack chose to go with CommonJS.

**Comment by ben-eb** [comment]
Going in another route, we could promote tools that instead of having to define an inline function for a config option, would require the module instead. So:

```js
module.exports = {
  foo: function (bar) {
    return bar + 'px';
  }
}
```

Would be this instead:

```json
{
  "foo": "./bar"
}
```

And `./bar.js`:

```js
module.exports = function (bar) {
  return bar + 'px';
}
```

Note that `./bar` could also be `bar` (in `node_modules`). Then, plugin configuration works across all formats.

**Comment by corysimmons** [comment]
I'm actually working on a project that fixes this exact problem so I think it's a non-issue. If someone comes to us with "How can I pass a function to a config option via CLI?" I will have a good answer for them.

My project is in a bad place right this second (I'm new to Node and making CLI's), but I'll try to get it all cleaned up this weekend.

**Comment by ben-eb** [comment]
We could require plugin authors to provide a string which resolved to a function, for those who are using functions as options.

**Comment by hawkrives** [comment]
(I'll just leave this here: technically, the js-yaml parser supports [loading and saving of JS functions](http://nodeca.github.io/js-yaml/#yaml=ZnVuYzogISFqcy9mdW5jdGlvbiA+CiAgZnVuY3Rpb24gZm9vYmFyKCkgewogICAgcmV0dXJuICdXb3chIEpTLVlBTUwgUm9ja3MhJzsKICB9Cg==))

**Comment by davidtheclark** [comment]
Brainstorm away --- Here's my last point, though: there is a clear and
simple solution in using a CommonJS config module like webpack's. It will
address everything we need, provide a single straightforward means of
configuration for all use-cases, and make this CLI's codebase very simple.
Given such a straightforward solution, I still think that the proposed
alternatives -- promoting specific tools at the expense of others, making
new demands of plugin authors, using weird js-yaml function syntax,
building an elaborate workaround -- are all mistakes that will add
confusion and weaken the effectiveness of this CLI in exchange for
relatively unimportant flexibility.

**Comment by ben-eb** [comment]
Just trying to make sense of the fact that other tools provide multiple configuration options. I am guessing that this is the case because config is inherently a personal thing. Perhaps when using PostCSS in a Ruby project, one might want their config to be consistently written in YAML, without one outlier in CJS.

**Comment by hawkrives** [comment]
Fair enough. I so feel, though, that if we document that, hey, you can configure all of the arguments to plugins *except for* the ones that take a function, then

1. I think that people can understand that.
2. Sometimes, you just want to try **something** without setting up an entire config, and I find that really annoying to do with Webpack. I usually end up falling back to Browserify, actually, which lets you configure far more from the command line.

**Comment by corysimmons** [comment]
@davidtheclark 99% of the time, people _will_ be using configs instead of trying to config everything from the CLI. No one is saying we shouldn't heavily promote this approach. I'm just suggesting that we try to make the CLI as flexible as possible so we can support as many use cases as possible.

No CLI supports passing functions as args. People still make CLI's that accept args...

> Sometimes, you just want to try something without setting up an entire config, and I find that really annoying to do with Webpack. I usually end up falling back to Browserify, actually, which lets you configure far more from the command line.

This a million times over. I don't want to configure something to play around with something... Anytime I make a semi-serious project (something I'm gonna be working on for more than 30 minutes) I configure it. Otherwise forcing people to configure something is a PITA.

**Comment by Siilwyn** [comment]
Finally had the time to read through this all. I think @davidtheclark brings the most sensible valid points to the table. Loading the configuration in the form of JavaScript enables for powerful options that can not be done in JSON on top of that this will work with all plugins.

We shouldn't reinvent the CLI but users should be able to configure the plugins by only using the CLI as far as that is possible in my opinion.

Also if we're passing multiple plugins with the same configuration flags wouldn't that cause issues? And thus mean we need to get the flags in a 'nested' form to prevent this problem?

**Comment by corysimmons** [comment]
> We shouldn't reinvent the CLI but users should be able to configure the plugins by only using the CLI as far as that is possible in my opinion.

I think everyone except @davidtheclark agrees with this.

> Also if we're passing multiple plugins with the same configuration flags wouldn't that cause issues? And thus mean we need to get the flags in a 'nested' form to prevent this problem?

@hawkrives @ben-eb I think we're leaning towards minimist. Is this true for minimist?

**Comment by hawkrives** [comment]
Like, `postcss -p autoprefixer --browsers '> 1%' -p cssnano ---browsers '> 1%'`?

Stock `minimist` gives us this, because it groups same-named arguments into arrays.

```json
{
  "_": ["input.css"],
  "p": [
    "autoprefixer",
    "cssnano"
  ],
  "browsers": [
    "> 1%",
    "> 1%"
  ],
  "o": "output.css"
}
```

This is kinda what `subarg` is designed for:

`postcss -p [autoprefixer --browsers '> 1%'] -p [cssnano --browsers '> 1%'] input.css -o output.css`

The brackets give let it know what key to associate the extra options with.

```json
{
  "_": ["input.css"],
  "p": [
    {
      "_": ["autoprefixer"],
      "browsers": "> 1%"
    },
    {
      "_": ["cssnano"],
      "browsers": "> 1%"
    }
  ],
  "o": "output.css"
}
```

Does that answer your question, @corysimmons?

**Comment by corysimmons** [comment]
Yeah thanks @hawkrives 

I agree with @ben-eb that minimist looks best, but minimist just doesn't seem capable of passing plugin options in a robust way.

Maybe we should go back to not worrying about what tool we're gonna use and just focus on the API. Once we get that we can make a tool.

So can we all agree on this syntax?

```fish
postcss -p autoprefixer --browsers '> 1%' --no-remove -p cssnano --browsers '> 1%' in.css > out.css
```

If so, I think we can fix minimist to match

```
{
  "_": ["input.css"],
  "p": [
    "autoprefixer", <--- this
    "cssnano" <--- that
  ],
  "browsers": [
    "> 1%", <--- this
    "> 1%" <--- that
  ],
  "o": "output.css"
}
```

**Comment by hawkrives** [comment]
Minimist looks nicer, I'll agree, but I really like the unambiguousness of the Subarg syntax.

**Comment by corysimmons** [comment]
lol, dammit, now I'm back to thinking `postcss -p foo [--option1 'foo' --option2 'bar']` is the best syntax... :(

**Comment by corysimmons** [comment]
We're back to https://github.com/postcss/postcss-cli/issues/1#issuecomment-178933137

@ai Read over this thread and pick one so we can move forward. :dancer:

**Comment by ai** [comment]
@corysimmons ha-ha :) this question is too tricky. Right now I am thinking too that `-p foo [--option1 'foo' --option2 'bar']` is the best. But it is soooo wierd.

**Comment by corysimmons** [comment]
@sindresorhus thinks we should:

`postcss --plugin="autoprefixer:{browsers: 'last 2', remove: 'false'}"`

https://github.com/sindresorhus/meow/issues/30#issuecomment-177828711

Not sure how I feel about it. On one hand, json-like syntax is pretty easy to parse. On the other hand, where does it stop? Do 2 plugins look like: `postcss --plugins="autoprefixer:{browsers: 'last 2', remove: 'false'}, lost:{flexbox 'flex', gutter: '60px'}"`

Anyone like that?

**Comment by sindresorhus** [comment]
> Do 2 plugins look like: postcss --plugins="autoprefixer:{browsers: 'last 2', remove: 'false'}, lost:{flexbox 'flex', gutter: '60px'}"

No:

```
postcss --plugin="autoprefixer:{browsers: 'last 2', remove: 'false'}" --plugin="something:{foo: 'bar'}"
```

-

Alternatively:

```
postcss --plugin-autoprefixer="browsers: 'last 2', remove: false" --plugin-something="foo: 'bar'"
```

Which I think I prefer.

**Comment by vitorgalvao** [comment]
I’m in agreement with @sindresorhus, for one simple important reason: clarity. I remember fighting with a tool (perhaps it was `postcss-cli`, even) some time ago while trying to figure out how to convert the examples in the a plugin pages to actual `--` commands (don’t remember the actual issue, perhaps there were nested commands, or something).

I’m a heavy CLI user and build a lot of (mostly) `bash` and `ruby` scripts with flags. Even though `--` flags makes sense and is familiar, in this case we’re passing options to other tools, and at some point it just becomes confusing.

Take [this autoprefixer example](https://github.com/postcss/autoprefixer#options). With @sindresorhus’ suggestion, it’d be trivial to convert. No need to transform their options into `--` flags, just copy and paste the example as is in the correct place.

From the two examples, I think I prefer the first, but any of them seems like an improvement over the alternatives.

**Comment by corysimmons** [comment]
`postcss --plugin-autoprefixer="browsers: 'last 2', remove: false" --plugin-something="foo: 'bar'"` I like that.

@ai Objections?

Also, is anyone considering stepping up to dev this?

**Comment by sindresorhus** [comment]
And as commented in https://github.com/sindresorhus/meow/issues/30#issuecomment-177828711, you can use [`levn`](https://github.com/gkz/levn) to parse the value.

**Comment by rafaelrinaldi** [comment]
Thinking about submitting a PR with improved instructions on the use of a configuration JSON (instead of CLI arguments). It was kinda tricky for me at first and some people at work also got confused by it.

@ai Thoughts?

**Comment by corysimmons** [comment]
It'd be nice if we exposed config files to all 3 standard config types as well as CLI args: https://github.com/davidtheclark/cosmiconfig

**Comment by corysimmons** [comment]
@ai Can you finalize this so someone can move forward?

**Comment by RyanZim** [comment]
@ai ping?

**Comment by ai** [comment]
@RyanZim I don’t use this project, so my thoughts will not be useful :)

**Comment by ai** [comment]
But, right now I am thinking about having one common config for any PostCSS runner: https://github.com/michael-ciniawsky/postcss-load-plugins and https://github.com/michael-ciniawsky/postcss-load-config

I think CLI also should use that common config.

**Comment by watilde** [comment]
Personally, I agreed on this https://github.com/postcss/postcss-cli/issues/1#issuecomment-245912873 and also I'd like to update all the cli option as well at the same time, like a breaking change.

**Comment by corysimmons** [comment]
👍 lgtm. Anyone want to volunteer as tribute?

**Comment by michael-ciniawsky** [comment]
👋 

``` --env|e ```

```
postcss --env|e  production
```
```js
{
  "name": "css",
  "main": "postcss.config.js",
  "scripts": {
    "css:prod": "NODE_ENV=production postcss -o dest/index.css src/index.css",
    "css:dev": "NODE_ENV=development postcss -o dest/index.css src/index.css",
  },
}
```

``` --help|h ```

```
postcss --help|h  $plugin
```

```npm home $plugin``` > ```npm repo $plugin``` > ```README.md``` (e.g github-man)

```--bundle|b ```
```
postcss -p sugarss -u [postcss-import --option foo --option bar] ...
```
```
postcss --bundle|b $name
```
**postcss-config-boilerplate**
```
|–index.js (postcss.config.js)
|-package.json
|-README.md
```
**package.json**
```js
{
  "name": "postcss-config-[name]",
  "main": "index.js",
  "postcss": {
    "parser": "sugarss",
    "plugins": {
      "postcss-import": { option: 'foo', option: 'bar' }
    }
  },
  "dependencies": {
     "postcss-import": "^8.1.2"
  }
}
```
**index.js (postcss.config.js)**
```js
const options = require('package.json').postcss
const plugins = require('package.json').postcss.plugins

module.exports = (ctx) => {
  parser: ctx.parser || options.parser
  plugins: {
     'postcss-import': ctx.import || plugins['postcss-import']
     ...
  }
}
```

**Comment by RyanZim** [comment]
@thomasklein Right now, we are working on postcss-cli v3. That is a complete rewrite that will use https://github.com/michael-ciniawsky/postcss-load-config, which uses cosmiconfig under the hood. See https://github.com/postcss/postcss-cli/projects/1 for more details and progress updates.

**Comment by RyanZim** [comment]
v3.0.0 is out, so going to say that fixes this issue. https://github.com/postcss/postcss-cli/releases/tag/v3.0.0

Please open new issues for anything that you think could be improved. Thanks for all the brainstorming here!
